### PR TITLE
[feat] Add support for evaluation dataset in `SFTTrainer`

### PR DESCRIPTION
**Pull request by SumanthRH**
# What does this PR do?

## Summary 

Adds evaluation-dataset support to `SFTTrainer`. During training, eval is run every `eval_interval` steps on a held-out dataset, and `eval_loss` is logged to wandb alongside `train_loss`. Built on top of [PR 1 (`pr1-forward-loss`)](../pr1-forward-loss/pr.md), which provides the Tinker-compatible `forward(loss_fn="cross_entropy", ...)` API + `WorkerOutput` dataclass that this PR consumes.

### `SFTConfig` fields

```python
eval_dataset_name: Optional[str] = None        # HF dataset name/path; None disables eval
eval_dataset_split: str = "validation"
eval_interval: int = 0                          # Run eval every N training steps; 0 disables
eval_before_train: bool = False        # Run evaluation at step 0 before training starts
```

Removes `eval_batch_size` — eval automatically uses `micro_train_batch_size_per_gpu * dp_size` per dispatch (via the new public `WorkerDispatch.dp_size(model)` accessor). Side effect: `eval_loss` is now the true per-non-pad-token NLL independent of batch size.

`collate_batch.batch_size` is now a required positional arg (was optional with a default fall-through to `sft_cfg.batch_size`).

### Eval loop in `SFTTrainer`

New `SFTTrainer.run_eval(eval_tokenized)`:
- Iterates the eval dataset in chunks of `micro_train_batch_size_per_gpu * dp_size`
- Calls `dispatch.forward("policy", batch, loss_fn="cross_entropy", loss_fn_config=None)` per chunk
- Aggregates a token-weighted mean loss across batches: `sum(batch_loss * nonpad_tokens) / sum(nonpad_tokens)`
- Returns `{"eval_loss": ...}`

### Example script

`examples/train/sft/run_sft_megatron_tulu3_eval.sh` — end-to-end SFT run on the tulu-3-sft-mixture dataset with the Megatron backend and eval enabled.

## Test plan

- [x] `tests/train/test_sft_tokenization.py` — 23 passed
- [x] End-to-end tulu Megatron run (Qwen2.5-0.5B-Instruct, 200 steps, eval every 50 steps) — eval_loss decreases monotonically: 2.1463 → 2.1419 → 2.1407 → 2.1400 ([wandb run 8t8opq7z](https://wandb.ai/sky-posttraining-uc-berkeley/skyrl-sft-eval/runs/8t8opq7z))
- [x] Verified wandb step-ordering bug is fixed: zero "Tried to log to step N that is less than the current step" warnings in the run log

## Dependencies

- Built on top of `pr1-forward-loss` — that PR must merge first.

### CODE REVIEW

File: skyrl/train/sft_trainer.py [hunk 1]
```diff
@@ -603,6 +615,76 @@ def load_checkpoint(self) -> int:
...
+        # Drop a trailing partial batch to keep DP/microbatch divisibility identical
+        # to training. Eval datasets are typically large enough that this is a no-op.
+        num_full_batches = num_eval // eval_chunk_size
+        if num_full_batches == 0:
+            raise ValueError(
+                f"Eval dataset has {num_eval} examples, which is less than the eval batch size "
+                f"of {eval_chunk_size} (micro_train_batch_size_per_gpu="
+                f"{self.sft_cfg.micro_train_batch_size_per_gpu} * dp_size={dp_size}). "
+                f"Use a larger eval split or reduce micro_train_batch_size_per_gpu."
+            )
```
[the remark this answers is not in the document]
Comment (SumanthRH) [reply skyrl/train/sft_trainer.py]: added!
